### PR TITLE
Merge duplicated code used in forall and exists.

### PR DESCRIPTION
**Pull request by hugoferreira**
A lot of code in `FluentCheckUniversal` and `FluentCheckExistential` was basically the same. Abstracted that away in `FluentCheckQuantifier`.

### CODE REVIEW

**Comment by arestivo** [comment]
Maybe use _s_ instead of _b_ as the parameter for breakCondition (as in _satisfiable_).

And maybe _check_ instead of _breakCondition_. 

Rationale: we are checking for satisfiability or non-satisfiability.

**Comment by hugoferreira** [comment]
Would that read well at the call site? We have:

```
if (this.breakCondition(result.satisfiable)) {
 return ...
}
```

Which makes sense... it is the condition that _stops_ the check, not the condition that _continues_  the step. Would you prefer a double negation, like, having this at the call site:

```
if (!this.check(result.satisfiable)) {
 return ...
}
```

... and changing the `forall` condition to:

```
check = (s => !s)
```

... it seems odd, right? We are checking for non-satisfiability in `forall`? And for satisfiability in `exists`? I really think that what we are doing is really early stopping.

**Comment by RuiGuedes** [comment]
I understand your points but to be honest it makes more sense to me to use @hugoferreira proposal because indeed we are really trying to early stop by shrinking. Also since there is already a `check()` function using check could lead to misinterpretations.

File: src/FluentCheck.ts [hunk 1]
```diff
@@ -175,58 +175,31 @@ class FluentCheckUniversal<K extends string, A, Rec extends ParentRec & Record<K
...
-    this.cache = this.dedup.sampleWithBias(this.configuration.sampleSize)
-  }
-
-  protected run(
-    testCase: FluentPicks,
-    callback: (arg: FluentPicks) => FluentResult,
-    partial: FluentResult | undefined = undefined,
-    depth = 0): FluentResult {
+  abstract defaultValue: boolean
+  abstract breakCondition: (b: boolean) => boolean
```
Comment (ruippeixotog) [inline src/FluentCheck.ts:188]: It looks simpler if we just have a `breakValue: boolean` here and do `result.satisfiable === breakValue`.

**Comment by hugoferreira** [comment]
Done. The only redundant thing I now see is that the initial condition is the negation of the break condition. You guys wanna leave this as it is, or make that unification?

**Comment by hugoferreira** [comment]
I also deleted the `initialValue` as it can be correctly derived from the break condition.

**Review by RuiGuedes — APPROVED** [review]

File: src/FluentCheck.ts [hunk 2]
```diff
@@ -175,58 +175,26 @@ class FluentCheckUniversal<K extends string, A, Rec extends ParentRec & Record<K
...
-      testCase[this.name] = tp
-      const result = callback(testCase)
-      if (result.satisfiable) {
-        result.addExample(this.name, tp)
-        return this.run(testCase, callback, result, depth + 1)
-      }
-    }
+class FluentCheckUniversal<K extends string, A, Rec extends ParentRec & Record<K, A>, ParentRec extends {}>
+  extends FluentCheckQuantifier<K, A, Rec, ParentRec> {
+  breakCondition = false
```
Comment (ruippeixotog) [inline src/FluentCheck.ts:192]: Now that it's a boolean I prefer if it is called `breakValue` as suggested rather than `breakCondition`.

**Review by ruippeixotog — CHANGES_REQUESTED** [review]

**Review by arestivo — APPROVED** [review]
Still not completely happy with the `breakValue` thingy, but I'll allow it.

**Pull request merged**